communication/include/communication/StateControl.hpp

```diff
@@ -14,31 +14,89 @@
 #include <functional>
 #include <utility>
 #include "callback_registry.hpp"
+#include <unordered_map>
+#include <variant>
+#include <memory>
+#include "Logger.hpp"
+
 namespace bft::communication {
+struct EnumHash {
+  template <typename T>
+  size_t operator()(T t) const {
+    return static_cast<size_t>(t);
+  }
+};
+
 class StateControl {
  public:
+  using CallbackRegistry = concord::util::CallbackRegistry<uint32_t>;
+  enum class EventType { TLS_COMM, THIN_REPLICA_SERVER };
+
   static StateControl& instance() {
     static StateControl instance_;
     return instance_;
   }
+
   void lockComm() { lock_comm_.lock(); }
   bool tryLockComm() { return lock_comm_.try_lock(); }
   void unlockComm() { lock_comm_.unlock(); }
 
-  void setCommRestartCallBack(std::function<void(uint32_t)> cb) {
-    if (cb != nullptr) comm_restart_cb_registry_.add(cb);
+  // Added/retained these methods to act as facade
+  void setCommRestartCallBack(std::function<void(uint32_t)> cb, const EventType& et = EventType::TLS_COMM) {
+    registerCallback(et, cb);
+  }
+
+  void restartComm(uint32_t id, const EventType& et = EventType::TLS_COMM) { invokeCallback(et, id); }
+
+  void setTlsRestartCallBack(std::function<void(uint32_t)> cb, const EventType& et = EventType::THIN_REPLICA_SERVER) {
+    registerCallback(et, cb);
   }
 
-  void restartComm(uint32_t id) { comm_restart_cb_registry_.invokeAll(id); }
+  void restartThinReplicaServer(uint32_t id, const EventType& et = EventType::THIN_REPLICA_SERVER) {
+    invokeCallback(et, id);
+  }
+
+  const std::string getEventTypeAsString(const EventType& et) {
+    switch (et) {
+      case EventType::TLS_COMM:
+        return "TLS_COMM";
+      case EventType::THIN_REPLICA_SERVER:
+        return "THIN_REPLICA_SERVER";
+      default:
+        return "Invalid Type";
+    }
+  }
   void setGetPeerPubKeyMethod(std::function<std::string(uint32_t)> m) { get_peer_pub_key_ = std::move(m); }
+
   std::string getPeerPubKey(uint32_t id) {
     if (get_peer_pub_key_) return get_peer_pub_key_(id);
     return std::string();
   }
 
  private:
+  StateControl() : logger_(logging::getLogger("concord-bft.comm.StetControl")) {}
+
+  void registerCallback(const EventType& et, std::function<void(uint32_t)> cb) {
+    if (cb != nullptr) {
+      if (event_registry_.find(et) == event_registry_.end()) {
+        event_registry_.insert({et, std::make_unique<CallbackRegistry>()});
+      }
+      event_registry_[et]->add(cb);
+    }
+  }
+
+  void invokeCallback(const EventType& et, uint32_t id) {
+    if (event_registry_.find(et) != event_registry_.end()) {
+      event_registry_[et]->invokeAll(id);
+    } else {
+      LOG_WARN(logger_, "No callback(s) registered for eventType " << getEventTypeAsString(et));
```

Comment:
Can this happen? and if yes (even in rare configurations?) you should think how often this log will print, might it produce too many entries?

Reply:
In live scenario, else case won't run. I had to add it for the current apollo test case. In that test case, client key exchange is triggered, hence tjis callback is called, since the callback is registered in ThinReplicaServer which is not in scope of this test case. Hence, I have added as WARN. but if u suggest, i can remove the else case itself.